[AI Accelerators] softmax kernel for Nested Tensor (CPU)

zrphercule opened the pull request:
Summary: Impl better softmax kernel for Nested Tensor CPU.

Test Plan:
Benchmark results:

On CPU (command: buck run mode/opt -c fbcode.platform=platform009 //pytext/fb/tools:benchmark_transformers  -- transformer --large --use-trt-kernel False --batch-size 16 --avg-sequence-length 64 --max-sequence-length 256 --iters 10 --use-real-data-distribution --module native --use-nt True --use-cpu True

With mask (previous impl):
NT: 4573.14 ms/iter, 0.14 TFLOP/s, Speedup: 2.33x;

Without mask:

NT: 3530.55 ms/iter, 0.18 TFLOP/s, Speedup: 1.51x

Reviewed By: mikekgfb

Differential Revision: D35679352

erichan1 commented on aten/src/ATen/native/transformers/attention.cpp:
```diff
@@ -115,27 +115,24 @@ Tensor bmm_nt(const Tensor& a, const Tensor& b) {
   return c_.view({a.size(0), a.size(1), a.size(2), b.size(2)});
 }
 
+void masked_softmax_dropout_cpu(
```
Where is this function used?

zrphercule replied:
Good point. I dont think this is used anywhere, should be just redundant stuff when resolving conflict, will remove it.

erichan1 reviewed (approved):
Thanks for adding the fast softmax kernel on CPU! Overall LGTM.

zrphercule commented:
@pytorchbot  merge -g

zrphercule commented:
@pytorchbot merge -g

zrphercule2 committed:
[AI Accelerators] softmax kernel for Nested Tensor (CPU) (#79756)

Summary:
Impl better softmax kernel for Nested Tensor CPU.

Pull Request resolved: https://github.com/pytorch/pytorch/pull/79756

Test Plan:
Benchmark results:

On CPU (command: buck run mode/opt -c fbcode.platform=platform009 //pytext/fb/tools:benchmark_transformers  -- transformer --large --use-trt-kernel False --batch-size 16 --avg-sequence-length 64 --max-sequence-length 256 --iters 10 --use-real-data-distribution --module native --use-nt True --use-cpu True

With mask (previous impl):
NT: 4573.14 ms/iter, 0.14 TFLOP/s, Speedup: 2.33x;

Without mask:

NT: 3530.55 ms/iter, 0.18 TFLOP/s, Speedup: 1.51x

Differential Revision: D35679352

Pulled By: zrphercule

fbshipit-source-id: 4d03ff53c88006185c5f5923682deb1fca778d11

zrphercule commented:
@pytorchbot  merge -g
